src/algebra/module/zlattice.lean

```diff
@@ -0,0 +1,236 @@
+/-
+Copyright (c) 2023 Xavier Roblot. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Xavier Roblot
+-/
+import analysis.normed.order.basic
+import measure_theory.group.fundamental_domain
+
+/-!
+# ℤ-lattices
+
+Let `E` be a finite dimensional vector space over a `normed_lattice_field` `K` that is also
+a `floor_ring`, e.g. `ℚ` or `ℝ`. A (full) ℤ-lattice `L` of `E` is a discrete subgroup of `E` such
+that `L` spans `E` over `K`.
+
+The ℤ-lattice `L` can be defined in two ways:
+* For `b` a basis of `E`, then `L : submodule.span ℤ (set.range b)` is a ℤ-lattice of `E`.
+* As `L : add_subgroup E` with the additional properties:
+  `(hd : ∀ r : ℝ, (L ∩ (metric.closed_ball 0 r)).finite)`, that is `L` is discrete
+  `(hs : submodule.span ℝ (L : set E) = ⊤)`, that is `L` spans `E` over `K`.
+
+## Main result
+* `zspan.is_add_fundamental_domain`: for `L : submodule.span ℤ (set.range b)` a ℤ-lattice, proves
+ that the set defined by `zspan.fundamental_domain` is indeed a fundamental domain.
+
+-/
+
+open_locale classical
+
+noncomputable theory
+
+section zspan
+
+open measure_theory measurable_set submodule
+
+variables {E ι : Type*}
+
+section basis
+
+variables {R K : Type*} [comm_ring R] [division_ring K] [algebra R K] [add_comm_group E]
+  [module K E] [module R E] [is_scalar_tower R K E] [no_zero_smul_divisors R K]
+variables (b : basis ι K E)
+
+variable (R)
+
+/-- The R-lattice spanned by `b` admits `b` as a R-basis. -/
+def zspan.basis : basis ι R (span R (set.range b)) :=
```

Comment:
This belongs in a more general file.

Reply:
All the results about restricting a `K`-basis to an `R`-basis have been generalized and moved to `linear_algebra.finite_dimensional`.